// @getpara/user-management-client/dist/esm/types/auth.js
export var EncryptorType;
(function (EncryptorType) {
    EncryptorType["USER"] = "USER";
    EncryptorType["RECOVERY"] = "RECOVERY";
    EncryptorType["BIOMETRICS"] = "BIOMETRICS";
    EncryptorType["PASSWORD"] = "PASSWORD";
})(EncryptorType || (EncryptorType = {}));
export var KeyShareType;
(function (KeyShareType) {
    KeyShareType["USER"] = "USER";
    KeyShareType["RECOVERY"] = "RECOVERY";
})(KeyShareType || (KeyShareType = {}));
export var PasswordStatus;
(function (PasswordStatus) {
    PasswordStatus["PENDING"] = "PENDING";
    PasswordStatus["COMPLETE"] = "COMPLETE";
})(PasswordStatus || (PasswordStatus = {}));
export var PublicKeyStatus;
(function (PublicKeyStatus) {
    PublicKeyStatus["PENDING"] = "PENDING";
    PublicKeyStatus["COMPLETE"] = "COMPLETE";
})(PublicKeyStatus || (PublicKeyStatus = {}));
export var PublicKeyType;
(function (PublicKeyType) {
    PublicKeyType["MOBILE"] = "MOBILE";
    PublicKeyType["WEB"] = "WEB";
})(PublicKeyType || (PublicKeyType = {}));
export var OAuthMethod;
(function (OAuthMethod) {
    OAuthMethod["GOOGLE"] = "GOOGLE";
    OAuthMethod["TWITTER"] = "TWITTER";
    OAuthMethod["APPLE"] = "APPLE";
    OAuthMethod["DISCORD"] = "DISCORD";
    OAuthMethod["FACEBOOK"] = "FACEBOOK";
    OAuthMethod["FARCASTER"] = "FARCASTER";
    OAuthMethod["TELEGRAM"] = "TELEGRAM";
})(OAuthMethod || (OAuthMethod = {}));
export var AuthMethod;
(function (AuthMethod) {
    AuthMethod["PASSWORD"] = "PASSWORD";
    AuthMethod["PASSKEY"] = "PASSKEY";
})(AuthMethod || (AuthMethod = {}));
