// @getpara/user-management-client/dist/esm/types/onRamp.js
export var OnRampProvider;
(function (OnRampProvider) {
    OnRampProvider["RAMP"] = "RAMP";
    OnRampProvider["STRIPE"] = "STRIPE";
    OnRampProvider["MOONPAY"] = "MOONPAY";
})(OnRampProvider || (OnRampProvider = {}));
export var OnRampAsset;
(function (OnRampAsset) {
    OnRampAsset["ETHEREUM"] = "ETHEREUM";
    OnRampAsset["USDC"] = "USDC";
    OnRampAsset["TETHER"] = "TETHER";
    OnRampAsset["POLYGON"] = "POLYGON";
    OnRampAsset["SOLANA"] = "SOLANA";
    OnRampAsset["ATOM"] = "ATOM";
    OnRampAsset["CELO"] = "CELO";
    OnRampAsset["CUSD"] = "CUSD";
    OnRampAsset["CEUR"] = "CEUR";
    OnRampAsset["CREAL"] = "CREAL";
})(OnRampAsset || (OnRampAsset = {}));
export var OnRampPurchaseStatus;
(function (OnRampPurchaseStatus) {
    OnRampPurchaseStatus["INITIATED"] = "INITIATED";
    OnRampPurchaseStatus["FINISHED"] = "FINISHED";
    OnRampPurchaseStatus["CANCELLED"] = "CANCELLED";
})(OnRampPurchaseStatus || (OnRampPurchaseStatus = {}));
export var OnRampPurchaseType;
(function (OnRampPurchaseType) {
    OnRampPurchaseType["BUY"] = "BUY";
    OnRampPurchaseType["SELL"] = "SELL";
})(OnRampPurchaseType || (OnRampPurchaseType = {}));
