// @getpara/user-management-client/dist/esm/types/wallet.js
import { OAuthMethod } from './auth.js';
export var WalletScheme;
(function (WalletScheme) {
    WalletScheme["DKLS"] = "DKLS";
    WalletScheme["CGGMP"] = "CGGMP";
    WalletScheme["ED25519"] = "ED25519";
})(WalletScheme || (WalletScheme = {}));
export var WalletType;
(function (WalletType) {
    WalletType["EVM"] = "EVM";
    WalletType["SOLANA"] = "SOLANA";
    WalletType["COSMOS"] = "COSMOS";
})(WalletType || (WalletType = {}));
export var Chain;
(function (Chain) {
    Chain["ETH"] = "ETH";
    Chain["CELO"] = "CELO";
    Chain["MATIC"] = "MATIC";
})(Chain || (Chain = {}));
export var Network;
(function (Network) {
    Network["ETHEREUM"] = "ETHEREUM";
    Network["SEPOLIA"] = "SEPOLIA";
    Network["ARBITRUM"] = "ARBITRUM";
    Network["BASE"] = "BASE";
    Network["OPTIMISM"] = "OPTIMISM";
    Network["POLYGON"] = "POLYGON";
    Network["SOLANA"] = "SOLANA";
    Network["COSMOS"] = "COSMOS";
    Network["CELO"] = "CELO";
    Network["NOBLE"] = "NOBLE";
})(Network || (Network = {}));
export const PREGEN_IDENTIFIER_TYPES = [
    'EMAIL',
    'PHONE',
    'CUSTOM_ID',
    OAuthMethod.DISCORD,
    OAuthMethod.TWITTER,
    OAuthMethod.TELEGRAM,
];
export const NON_ED25519 = [WalletScheme.DKLS, WalletScheme.CGGMP];
