// @getpara/core-sdk/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SHORT_POLLING_INTERVAL_MS = exports.POLLING_INTERVAL_MS = exports.SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR = exports.LOCAL_STORAGE_SESSION_COOKIE = exports.LOCAL_STORAGE_CURRENT_EXTERNAL_WALLET_ADDRESSES = exports.LOCAL_STORAGE_CURRENT_WALLET_IDS = exports.LOCAL_STORAGE_EXTERNAL_WALLETS = exports.LOCAL_STORAGE_WALLETS = exports.LOCAL_STORAGE_ED25519_WALLETS = exports.LOCAL_STORAGE_USER_ID = exports.LOCAL_STORAGE_TELEGRAM_USER_ID = exports.LOCAL_STORAGE_FARCASTER_USERNAME = exports.LOCAL_STORAGE_COUNTRY_CODE = exports.LOCAL_STORAGE_PHONE = exports.LOCAL_STORAGE_EMAIL = exports.PREFIX = exports.PARA_CORE_VERSION = void 0;
exports.PARA_CORE_VERSION = '1.1.0';
exports.PREFIX = '@CAPSULE/';
exports.LOCAL_STORAGE_EMAIL = `${exports.PREFIX}e-mail`;
exports.LOCAL_STORAGE_PHONE = `${exports.PREFIX}phone`;
exports.LOCAL_STORAGE_COUNTRY_CODE = `${exports.PREFIX}countryCode`;
exports.LOCAL_STORAGE_FARCASTER_USERNAME = `${exports.PREFIX}farcasterUsername`;
exports.LOCAL_STORAGE_TELEGRAM_USER_ID = `${exports.PREFIX}telegramUserId`;
exports.LOCAL_STORAGE_USER_ID = `${exports.PREFIX}userId`;
exports.LOCAL_STORAGE_ED25519_WALLETS = `${exports.PREFIX}ed25519Wallets`;
exports.LOCAL_STORAGE_WALLETS = `${exports.PREFIX}wallets`;
exports.LOCAL_STORAGE_EXTERNAL_WALLETS = `${exports.PREFIX}externalWallets`;
exports.LOCAL_STORAGE_CURRENT_WALLET_IDS = `${exports.PREFIX}currentWalletIds`;
exports.LOCAL_STORAGE_CURRENT_EXTERNAL_WALLET_ADDRESSES = `${exports.PREFIX}currentExternalWalletAddresses`;
exports.LOCAL_STORAGE_SESSION_COOKIE = `${exports.PREFIX}sessionCookie`;
exports.SESSION_STORAGE_LOGIN_ENCRYPTION_KEY_PAIR = `${exports.PREFIX}loginEncryptionKeyPair`;
exports.POLLING_INTERVAL_MS = 2000;
exports.SHORT_POLLING_INTERVAL_MS = 1000;
