// @getpara/core-sdk/dist/cjs/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransactionReviewTimeout = exports.TransactionReviewDenied = exports.TransactionReviewError = void 0;
class TransactionReviewError extends Error {
    constructor(transactionReviewUrl) {
        super('transaction review error');
        this.name = 'TransactionReviewError';
        this.transactionReviewUrl = transactionReviewUrl;
    }
}
exports.TransactionReviewError = TransactionReviewError;
class TransactionReviewDenied extends Error {
    constructor() {
        super('transaction review has been denied by the user');
        this.name = 'TransactionReviewDenied';
    }
}
exports.TransactionReviewDenied = TransactionReviewDenied;
class TransactionReviewTimeout extends Error {
    constructor(transactionReviewUrl, pendingTransactionId) {
        super('transaction review has timed out');
        this.name = 'TransactionReviewTimeout';
        this.transactionReviewUrl = transactionReviewUrl;
        this.pendingTransactionId = pendingTransactionId;
    }
}
exports.TransactionReviewTimeout = TransactionReviewTimeout;
