// @getpara/server-sdk/dist/esm/wallet/privateKey.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as uuid from 'uuid';
import { setupWorker } from '../workers/workerWrapper.js';
export function getPrivateKey(ctx, userId, walletId, share, sessionCookie) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
            const workId = uuid.v4();
            const worker = yield setupWorker(ctx, (res) => __awaiter(this, void 0, void 0, function* () {
                resolve(res);
            }), workId);
            worker.postMessage({
                env: ctx.env,
                apiKey: ctx.apiKey,
                cosmosPrefix: ctx.cosmosPrefix,
                params: { share, walletId, userId },
                functionType: 'GET_PRIVATE_KEY',
                offloadMPCComputationURL: ctx.offloadMPCComputationURL,
                disableWorkers: ctx.disableWorkers,
                sessionCookie,
                useDKLS: ctx.useDKLS,
                disableWebSockets: ctx.disableWebSockets,
                wasmOverride: ctx.wasmOverride,
                workId,
            });
        }));
    });
}
