// @getpara/server-sdk/dist/esm/wallet/signing.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as uuid from 'uuid';
import { setupWorker } from '../workers/workerWrapper.js';
export function signTransaction(ctx, userId, walletId, share, tx, chainId, sessionCookie, isDKLS) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
            const workId = uuid.v4();
            const worker = yield setupWorker(ctx, (sendTransactionRes) => __awaiter(this, void 0, void 0, function* () {
                resolve(sendTransactionRes);
            }), workId);
            worker.postMessage({
                env: ctx.env,
                apiKey: ctx.apiKey,
                cosmosPrefix: ctx.cosmosPrefix,
                params: { share, walletId, userId, tx, chainId },
                functionType: 'SIGN_TRANSACTION',
                offloadMPCComputationURL: ctx.offloadMPCComputationURL,
                disableWorkers: ctx.disableWorkers,
                sessionCookie,
                useDKLS: isDKLS,
                disableWebSockets: ctx.disableWebSockets,
                wasmOverride: ctx.wasmOverride,
                workId,
            });
        }));
    });
}
export function sendTransaction(ctx, userId, walletId, share, tx, chainId, sessionCookie, isDKLS) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
            const workId = uuid.v4();
            const worker = yield setupWorker(ctx, (sendTransactionRes) => __awaiter(this, void 0, void 0, function* () {
                resolve(sendTransactionRes);
            }), workId);
            worker.postMessage({
                env: ctx.env,
                apiKey: ctx.apiKey,
                params: { share, walletId, userId, tx, chainId },
                functionType: 'SEND_TRANSACTION',
                offloadMPCComputationURL: ctx.offloadMPCComputationURL,
                disableWorkers: ctx.disableWorkers,
                sessionCookie,
                useDKLS: isDKLS,
                disableWebSockets: ctx.disableWebSockets,
                wasmOverride: ctx.wasmOverride,
                workId,
            });
        }));
    });
}
export function signMessage(ctx, userId, walletId, share, message, sessionCookie, isDKLS) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
            const workId = uuid.v4();
            const worker = yield setupWorker(ctx, (signMessageRes) => __awaiter(this, void 0, void 0, function* () {
                resolve(signMessageRes);
            }), workId);
            worker.postMessage({
                env: ctx.env,
                apiKey: ctx.apiKey,
                params: { share, walletId, userId, message },
                functionType: 'SIGN_MESSAGE',
                offloadMPCComputationURL: ctx.offloadMPCComputationURL,
                disableWorkers: ctx.disableWorkers,
                sessionCookie,
                useDKLS: isDKLS,
                disableWebSockets: ctx.disableWebSockets,
                wasmOverride: ctx.wasmOverride,
                workId,
            });
        }));
    });
}
export function ed25519Sign(ctx, userId, walletId, share, base64Bytes, sessionCookie) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield new Promise((resolve) => __awaiter(this, void 0, void 0, function* () {
            const workId = uuid.v4();
            const worker = yield setupWorker(ctx, (signMessageRes) => __awaiter(this, void 0, void 0, function* () {
                resolve(signMessageRes);
            }), workId);
            worker.postMessage({
                env: ctx.env,
                apiKey: ctx.apiKey,
                params: { share, walletId, userId, base64Bytes },
                functionType: 'ED25519_SIGN',
                offloadMPCComputationURL: ctx.offloadMPCComputationURL,
                disableWorkers: ctx.disableWorkers,
                sessionCookie,
                disableWebSockets: ctx.disableWebSockets,
                wasmOverride: ctx.wasmOverride,
                workId,
            });
        }));
    });
}
