// @getpara/user-management-client/dist/cjs/types/auth.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthMethod = exports.OAuthMethod = exports.PublicKeyType = exports.PublicKeyStatus = exports.PasswordStatus = exports.KeyShareType = exports.EncryptorType = void 0;
var EncryptorType;
(function (EncryptorType) {
    EncryptorType["USER"] = "USER";
    EncryptorType["RECOVERY"] = "RECOVERY";
    EncryptorType["BIOMETRICS"] = "BIOMETRICS";
    EncryptorType["PASSWORD"] = "PASSWORD";
})(EncryptorType || (exports.EncryptorType = EncryptorType = {}));
var KeyShareType;
(function (KeyShareType) {
    KeyShareType["USER"] = "USER";
    KeyShareType["RECOVERY"] = "RECOVERY";
})(KeyShareType || (exports.KeyShareType = KeyShareType = {}));
var PasswordStatus;
(function (PasswordStatus) {
    PasswordStatus["PENDING"] = "PENDING";
    PasswordStatus["COMPLETE"] = "COMPLETE";
})(PasswordStatus || (exports.PasswordStatus = PasswordStatus = {}));
var PublicKeyStatus;
(function (PublicKeyStatus) {
    PublicKeyStatus["PENDING"] = "PENDING";
    PublicKeyStatus["COMPLETE"] = "COMPLETE";
})(PublicKeyStatus || (exports.PublicKeyStatus = PublicKeyStatus = {}));
var PublicKeyType;
(function (PublicKeyType) {
    PublicKeyType["MOBILE"] = "MOBILE";
    PublicKeyType["WEB"] = "WEB";
})(PublicKeyType || (exports.PublicKeyType = PublicKeyType = {}));
var OAuthMethod;
(function (OAuthMethod) {
    OAuthMethod["GOOGLE"] = "GOOGLE";
    OAuthMethod["TWITTER"] = "TWITTER";
    OAuthMethod["APPLE"] = "APPLE";
    OAuthMethod["DISCORD"] = "DISCORD";
    OAuthMethod["FACEBOOK"] = "FACEBOOK";
    OAuthMethod["FARCASTER"] = "FARCASTER";
    OAuthMethod["TELEGRAM"] = "TELEGRAM";
})(OAuthMethod || (exports.OAuthMethod = OAuthMethod = {}));
var AuthMethod;
(function (AuthMethod) {
    AuthMethod["PASSWORD"] = "PASSWORD";
    AuthMethod["PASSKEY"] = "PASSKEY";
})(AuthMethod || (exports.AuthMethod = AuthMethod = {}));
