// @getpara/user-management-client/dist/cjs/types/wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NON_ED25519 = exports.PREGEN_IDENTIFIER_TYPES = exports.Network = exports.Chain = exports.WalletType = exports.WalletScheme = void 0;
const auth_js_1 = require("./auth.js");
var WalletScheme;
(function (WalletScheme) {
    WalletScheme["DKLS"] = "DKLS";
    WalletScheme["CGGMP"] = "CGGMP";
    WalletScheme["ED25519"] = "ED25519";
})(WalletScheme || (exports.WalletScheme = WalletScheme = {}));
var WalletType;
(function (WalletType) {
    WalletType["EVM"] = "EVM";
    WalletType["SOLANA"] = "SOLANA";
    WalletType["COSMOS"] = "COSMOS";
})(WalletType || (exports.WalletType = WalletType = {}));
var Chain;
(function (Chain) {
    Chain["ETH"] = "ETH";
    Chain["CELO"] = "CELO";
    Chain["MATIC"] = "MATIC";
})(Chain || (exports.Chain = Chain = {}));
var Network;
(function (Network) {
    Network["ETHEREUM"] = "ETHEREUM";
    Network["SEPOLIA"] = "SEPOLIA";
    Network["ARBITRUM"] = "ARBITRUM";
    Network["BASE"] = "BASE";
    Network["OPTIMISM"] = "OPTIMISM";
    Network["POLYGON"] = "POLYGON";
    Network["SOLANA"] = "SOLANA";
    Network["COSMOS"] = "COSMOS";
    Network["CELO"] = "CELO";
    Network["NOBLE"] = "NOBLE";
})(Network || (exports.Network = Network = {}));
exports.PREGEN_IDENTIFIER_TYPES = [
    'EMAIL',
    'PHONE',
    'CUSTOM_ID',
    auth_js_1.OAuthMethod.DISCORD,
    auth_js_1.OAuthMethod.TWITTER,
    auth_js_1.OAuthMethod.TELEGRAM,
];
exports.NON_ED25519 = [WalletScheme.DKLS, WalletScheme.CGGMP];
