// @getpara/core-sdk/dist/cjs/external/mpcComputationClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initClient = void 0;
const axios_1 = __importDefault(require("axios"));
function initClient(baseURL, useAdapter) {
    const client = axios_1.default.create({ baseURL });
    if (useAdapter) {
        client.defaults.adapter = function (config) {
            return fetch(config.baseURL + config.url, {
                method: config.method,
                headers: config.headers,
                body: config.data,
                credentials: config.withCredentials ? 'include' : undefined,
            })
                .then(response => response.text().then(text => ({
                data: text,
                status: response.status,
                statusText: response.statusText,
                headers: response.headers,
                config: config,
                request: fetch,
            })))
                .catch(function (reason) {
                throw reason;
            });
        };
    }
    return client;
}
exports.initClient = initClient;
